Unexpected end of file from server

**Describe the bug**

>    Web service requests that are going trough proxy(IHS) with a low KeepAliveTimeout(3), fails with "java.net.SocketException: Unexpected end of file from server" for Liberty versions 2.0.0.8 and later.

Steps:
> 1. Requester sends a web service request
> 2. Keep alive timeout at the proxy is 3 seconds.
> 3. In 3 seconds, the proxy sends to a FIN packet to the requester.
> 4. If the requester sends another request just before the FIN packet, the request fails with "java.net.SocketException: Unexpected end of file from server"

**Topology:**
> WebService requester/wlp     ---     forward proxy(IHS)     ---     WebService provider/wlp

If there is a stack trace, please include the FULL stack trace (without any `[internal classes]` lines in it). To find the full stack trace, you may need to check in `$WLP_OUTPUT_DIR/messages.log`
```
Caused by: java.net.SocketException: Unexpected end of file from server
        at sun.net.www.http.HttpClient.parseHTTPHeader(HttpClient.java:884)
        at sun.net.www.http.HttpClient.parseHTTP(HttpClient.java:709)
        at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1619)
        at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1524)
        at java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:491)
        at org.apache.cxf.transport.http.URLConnectionHTTPConduit$URLConnectionWrappedOutputStream$2.run(URLConnectionHTTPConduit.java:424)
        at org.apache.cxf.transport.http.URLConnectionHTTPConduit$URLConnectionWrappedOutputStream$2.run(URLConnectionHTTPConduit.java:420)
        at java.security.AccessController.doPrivileged(AccessController.java:747)
        at org.apache.cxf.transport.http.URLConnectionHTTPConduit$URLConnectionWrappedOutputStream.getResponseCode(URLConnectionHTTPConduit.java:420)
        at org.apache.cxf.transport.http.HTTPConduit$WrappedOutputStream.doProcessResponseCode(HTTPConduit.java:1659)
        at org.apache.cxf.transport.http.HTTPConduit$WrappedOutputStream.handleResponseInternal(HTTPConduit.java:1687)
        at org.apache.cxf.transport.http.HTTPConduit$WrappedOutputStream.handleResponse(HTTPConduit.java:1632)
        at org.apache.cxf.transport.http.HTTPConduit$WrappedOutputStream.close(HTTPConduit.java:1424)
        ... 41 more
```

**Steps to Reproduce**  
> 1. Have Web Service consumer and provider on different Liberty servers 
> 2. Configure IHS(IBM HTTP Server) as a forward proxy. 
> 3. Set KeepAliveTimeout to 3
> 4. Set Listen port same as http.proxyPort at the client side
> 5. Start IHS and liberty
> 6. Open a new terminal window and run the tcpdump command.
> 7. Send a request curl to invoke Web Service consumer
> 8. After the request is completed, IHS will send back a FIN packet to client in 3 seconds.
> 9. Send another request by curl. This will fail on wlp 22.0.0.8 or later.

**Expected behavior**  
> Web Service should run trough proxy with success. 

**Diagnostic information:**  
>  - OpenLiberty Version: 22.0.0.8 and above
>  - Affected feature(s)  :  jaxws-2.2
>  - Java Version             :  Java(TM) SE Runtime Environment (8.0.6.5 - pap6480sr6fp5-20200111_02(SR6 FP5))
>  - server.xml configuration 
>  - If it would be useful, upload the messages.log file found in `$WLP_OUTPUT_DIR/messages.log`

**Additional context**

--- a/dev/com.ibm.ws.org.apache.cxf.cxf.rt.transports.http.3.2/src/org/apache/cxf/transport/http/URLConnectionHTTPConduit.java
+++ b/dev/com.ibm.ws.org.apache.cxf.cxf.rt.transports.http.3.2/src/org/apache/cxf/transport/http/URLConnectionHTTPConduit.java
@@ -472,6 +472,27 @@ public class URLConnectionHTTPConduit extends HTTPConduit {
         protected void setFixedLengthStreamingMode(int i) {
             // [CXF-6227] do not call connection.setFixedLengthStreamingMode(i)
             // to prevent https://bugs.openjdk.java.net/browse/JDK-8044726
+        
+            // Liberty change for issue #25866-unexpected_EOF_from_server
+            // The bug [CXF-6227] is a performance bug observed for Java 7
+            // In case of a performance drop is observed need to create a walk around
+            if(isTrue("cxf.set.fixedlengthstreaming"))        {
+                connection.setFixedLengthStreamingMode(i);
+                LOG.fine("~setFixedLengthStreamingMode is called");
+            } else {
+                LOG.fine("~setFixedLengthStreamingMode is NOT called");
+            }
+        }
+
+        // Liberty Change
+        private boolean isTrue(String propertyName)     {
+           if(propertyName != null && !propertyName.isEmpty()) {
+               String propertyValue = System.getProperty(propertyName);
+               if(propertyValue != null && propertyValue.equalsIgnoreCase("true"))      {
+                   return true;
+               }
+           }
+           return false;
         }
         @FFDCIgnore(PrivilegedActionException.class) // Liberty Change
         protected void handleNoOutput() throws IOException {

--- a/dev/com.ibm.ws.org.apache.cxf.cxf.rt.transports.http.3.2/src/org/apache/cxf/transport/http/URLConnectionHTTPConduit.java
+++ b/dev/com.ibm.ws.org.apache.cxf.cxf.rt.transports.http.3.2/src/org/apache/cxf/transport/http/URLConnectionHTTPConduit.java
@@ -473,18 +473,15 @@ public class URLConnectionHTTPConduit extends HTTPConduit {
             // [CXF-6227] do not call connection.setFixedLengthStreamingMode(i)
             // to prevent https://bugs.openjdk.java.net/browse/JDK-8044726
         
-            // Liberty change for issue #25866-unexpected_EOF_from_server
+            // Liberty Change for issue #25866-unexpected_EOF_from_server
             // The bug [CXF-6227] is a performance bug observed for Java 7
             // In case of a performance drop is observed need to create a walk around
             if(isTrue("cxf.set.fixedlengthstreaming"))        {
                 connection.setFixedLengthStreamingMode(i);
-                LOG.fine("~setFixedLengthStreamingMode is called");
-            } else {
-                LOG.fine("~setFixedLengthStreamingMode is NOT called");
-            }
+            } // Liberty Change End
         }
 
-        // Liberty Change
+        // Liberty Change Begin
         private boolean isTrue(String propertyName)     {
            if(propertyName != null && !propertyName.isEmpty()) {
                String propertyValue = System.getProperty(propertyName);
@@ -494,6 +491,7 @@ public class URLConnectionHTTPConduit extends HTTPConduit {
            }
            return false;
         }
+        // Liberty Change Begin
         @FFDCIgnore(PrivilegedActionException.class) // Liberty Change
         protected void handleNoOutput() throws IOException {
             if ("POST".equals(getMethod())) {

--- a/dev/com.ibm.ws.org.apache.cxf.cxf.rt.transports.http.3.2/src/org/apache/cxf/transport/http/URLConnectionHTTPConduit.java
+++ b/dev/com.ibm.ws.org.apache.cxf.cxf.rt.transports.http.3.2/src/org/apache/cxf/transport/http/URLConnectionHTTPConduit.java
@@ -33,6 +33,9 @@ import java.net.URLConnection;
 import java.security.AccessController;
 import java.security.PrivilegedActionException;
 import java.security.PrivilegedExceptionAction;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
 import java.util.logging.Level;
 
 import javax.net.ssl.HttpsURLConnection;
@@ -43,6 +46,7 @@ import org.apache.cxf.common.util.SystemPropertyAction;
 import org.apache.cxf.configuration.jsse.TLSClientParameters;
 import org.apache.cxf.helpers.IOUtils;
 import org.apache.cxf.io.CacheAndWriteOutputStream;
+import org.apache.cxf.message.Exchange;
 import org.apache.cxf.message.Message;
 import org.apache.cxf.message.MessageUtils;
 import org.apache.cxf.service.model.EndpointInfo;
@@ -475,23 +479,20 @@ public class URLConnectionHTTPConduit extends HTTPConduit {
         
             // Liberty Change for issue #25866-unexpected_EOF_from_server
             // The bug [CXF-6227] is a performance bug observed for Java 7
-            // In case of a performance drop is observed need to create a walk around
-            if(isTrue("cxf.set.fixedlengthstreaming"))        {
+            // In case of a performance drop is observed, need to create a work around
+            boolean isRestMessage = true; // It's safer to set isRestMessage, because 
+            // calling setFixedLengthStreamingMode method creates IllegalStateException at JAX-RS 
+            Exchange exchange = outMessage.getExchange();
+            if(exchange != null)        {
+                //This property return true if it's a rest web service
+                isRestMessage = PropertyUtils.isTrue(exchange.get(org.apache.cxf.message.Message.REST_MESSAGE));
+            }
+            // Call setFixedLengthStreamingMode for JAX-WS only
+            if(!isRestMessage)   { 
                 connection.setFixedLengthStreamingMode(i);
-            } // Liberty Change End
+            }
+            // Liberty Change End
         }
-
-        // Liberty Change Begin
-        private boolean isTrue(String propertyName)     {
-           if(propertyName != null && !propertyName.isEmpty()) {
-               String propertyValue = System.getProperty(propertyName);
-               if(propertyValue != null && propertyValue.equalsIgnoreCase("true"))      {
-                   return true;
-               }
-           }
-           return false;
-        }
-        // Liberty Change Begin
         @FFDCIgnore(PrivilegedActionException.class) // Liberty Change
         protected void handleNoOutput() throws IOException {
             if ("POST".equals(getMethod())) {

